### PR TITLE
[core] Don't use signed int type for anchor segment

### PR DESCRIPTION
**Pull request by alexshalamov**
Erroneous signed to unsigned implicit conversion was used  when PlacedSymbol(s) were created in SymbolLayout and signed values were used as indexes in a few other places.

### CODE REVIEW

File: src/mbgl/geometry/anchor.hpp [hunk 1]
```diff
@@ -11,9 +12,9 @@ class Anchor {
     Point<float> point;
     float angle = 0.0f;
     float scale = 0.0f;
-    int segment = -1;
+    optional<std::size_t> segment = nullopt;
```
Comment (pozdnyakov) [inline src/mbgl/geometry/anchor.hpp]: nit: just `optional<std::size_t> segment;`

File: src/mbgl/layout/symbol_layout.cpp [hunk 2]
```diff
@@ -695,18 +695,20 @@ bool SymbolLayout::anchorIsTooClose(const std::u16string& text, const float repe
...
-    if (anchor.segment != -1) {
-        auto sumForwardLength = util::dist<float>(anchor.point, line[anchor.segment + 1]);
-        auto sumBackwardLength = util::dist<float>(anchor.point, line[anchor.segment]);
-        for (size_t i = anchor.segment + 1; i < line.size(); i++) {
+    if (anchor.segment) {
+        std::size_t segment = *anchor.segment;
+        assert(segment < line.size());
+        auto sumForwardLength = util::dist<float>(anchor.point, line[segment + 1]);
+        auto sumBackwardLength = util::dist<float>(anchor.point, line[segment]);
+        for (std::size_t i = segment + 1; i < line.size(); i++) {
```
Comment (pozdnyakov) [inline src/mbgl/layout/symbol_layout.cpp]: nit: `++i`

File: src/mbgl/layout/symbol_layout.cpp [hunk 3]
```diff
@@ -695,18 +695,20 @@ bool SymbolLayout::anchorIsTooClose(const std::u16string& text, const float repe
...
             tileDistances[i] = sumForwardLength;
             if (i < line.size() - 1) {
                 sumForwardLength += util::dist<float>(line[i + 1], line[i]);
             }
         }
-        for (auto i = anchor.segment; i >= 0; i--) {
+        for (std::size_t i = ++segment; i-- > 0;) {
```
Comment (pozdnyakov) [inline src/mbgl/layout/symbol_layout.cpp]: `--i != 0u`

Comment (pozdnyakov) [reply src/mbgl/layout/symbol_layout.cpp]: `std::size_t i = segment + 1`

File: src/mbgl/layout/symbol_layout.cpp [hunk 4]
```diff
@@ -695,18 +695,20 @@ bool SymbolLayout::anchorIsTooClose(const std::u16string& text, const float repe
...
             tileDistances[i] = sumBackwardLength;
-            if (i > 0) {
+            if (i != 0u) {
```
Comment (pozdnyakov) [inline src/mbgl/layout/symbol_layout.cpp]: no need to check here, could be checked in the cycle conditions

File: src/mbgl/text/collision_feature.cpp [hunk 5]
```diff
@@ -82,16 +86,14 @@ void CollisionFeature::bboxifyLabel(const GeometryCoordinates& line, GeometryCoo
...
     float anchorDistance = firstBoxOffset;
     const float labelStartDistance = -labelLength / 2;
     const float paddingStartDistance = labelStartDistance - labelLength / 8;
 
     // move backwards along the line to the first segment the label appears on
     do {
-        index--;
-
-        if (index < 0) {
+        if (index == 0u) {
```
Comment (pozdnyakov) [inline src/mbgl/text/collision_feature.cpp]: this whole condition does not seem needed, could be:

```
do {
} while (anchorDistance > paddingStartDistance && --index != 0u)
```

File: src/mbgl/layout/symbol_layout.cpp (hunk 3 above)
Comment (alexshalamov) [reply src/mbgl/layout/symbol_layout.cpp]: Original algorithm included 0th index, thus, I can't use `!= 0u` condition.

Comment (pozdnyakov) [reply src/mbgl/layout/symbol_layout.cpp]: :+1:

File: src/mbgl/text/collision_feature.cpp (hunk 5 above)
Comment (alexshalamov) [reply src/mbgl/text/collision_feature.cpp]: I don't think that is correct, `--index != 0u` will short circuit loop early and condition, e.g., `anchorDistance > labelStartDistance)` will never be reachable.

**Pull request merged**